librz/util/graph.c

```diff
@@ -316,3 +316,893 @@ RZ_API void rz_graph_dfs(RzGraph *g, RzGraphVisitor *vis) {
 		free(color);
 	}
 }
+
+#include <rz_util/rz_graph_new.h>
+
+typedef struct rz_graph_list_edge_impl_t {
+	HtUP *adjacent_table; // <ut64, RzList *>
+} RzGraphListEdgeImpl;
+
+typedef struct rz_graph_matrix_edge_impl_t {
+	st64 **adj_matrix;
+	unsigned int n_nodes;
+} RzGraphMatrixEdgeImpl;
+
+/**
+ * @brief Adds an edge between the nodes `from` and `to` in the graph.
+ * @param graph Listed Implementation of the graph.l
+ * @param from source node.
+ * @param to destination node.
+ * @return true if the edge is added successfully, false otherwise.
+ */
+static bool list_impl_add_edge(RzGraphNew *graph, const RzGraphNodeNew *from, const RzGraphNodeNew *to) {
+	if (!graph || !from || !to) {
+		rz_return_val_if_fail(graph && from && to, false);
+		return false;
+	}
+
+	RzGraphListEdgeImpl *impl = (RzGraphListEdgeImpl *)graph->edge_set->impl;
+	if (!impl) {
+		RZ_LOG_WARN("No implementation found for the graph")
+		return false;
+	}
+
+	bool from_found = false, to_found = false;
+	ht_up_find(graph->nodes, from->hash_id, &from_found);
+	ht_up_find(graph->nodes, to->hash_id, &to_found);
+	if (!from_found || !to_found) {
+		RZ_LOG_WARN("Node not found in the graph")
+		return false;
+	}
+
+	from_found = false;
+	RzList *neighbours = ht_up_find(impl->adjacent_table, from->hash_id, &from_found);
+
+	if (!from_found) {
+		neighbours = rz_list_new();
+		if (!neighbours) {
+			RZ_LOG_WARN("Failed to allocate memory for the neighbours list")
+			return false;
+		}
+		rz_list_append(neighbours, to);
+		ht_up_insert(impl->adjacent_table, from->hash_id, neighbours);
+		return true;
+	}
+
+	to_found = false;
+	rz_list_find(neighbours, to, NULL, &to_found);
+	if (to_found) {
+		RZ_LOG_INFO("An edge already exists between the nodes")
+		return false;
+	}
+
+	rz_list_append(neighbours, to);
+	return true;
+}
+
+/**
+ * @brief Removes an edge between the nodes `from` and `to` in the graph.
+ * @param graph Listed Implementation of the graph.
+ * @param from source node
+ * @param to end node
+ * @return true if the edge is removed successfully, false otherwise.
+ */
+static bool list_impl_remove_edge(RzGraphNew *graph, const RzGraphNodeNew *from, const RzGraphNodeNew *to) {
+	if (!graph || !from || !to) {
+		rz_return_val_if_fail(graph && from && to, false);
+		return false;
+	}
+
+	RzGraphListEdgeImpl *impl = (RzGraphListEdgeImpl *)graph->edge_set->impl;
+	if (!impl) {
+		RZ_LOG_WARN("No implementation found for the graph")
+		return false;
+	}
+
+	bool from_found = false, to_found = false;
+	ht_up_find(graph->nodes, from->hash_id, &from_found);
+	ht_up_find(graph->nodes, to->hash_id, &to_found);
+	if (!from_found || !to_found) {
+		RZ_LOG_WARN("Node not found in the graph")
```

Comment:
```suggestion
		RZ_LOG_WARN("Node not found in the graph.\n");
```

The new lines are missing.